Lms/admin diagnostic csv download

anathomical opened the pull request:
## WHAT
- Update Admin Diagnostic Report UI to add Download buttons (including one on the Students tab when a user has "loaded more" enough times to show 500 students, the current cap on that page)
- Add a new API controller action that allows a user to request download links for CSVs of their report
- Add a worker to extract the user's saved filters so that the CSV filters the same way that the user's web reports do
- Add a worker to run all the Admin Diagnostic queries, roll up the data, put it into CSVs, stash those CSVs in S3, and then send an email to the user with links to those CSVs
- Add adapters that can convert the results of Admin Diagnostic queries into CSVs
- Add a Mailer to assemble the email with links to the CSVs to be sent to requesting users
## WHY
We want admins to be able to take their Diagnostic data and plug it into their own data analysis tools for maximum flexibility.  We also want to allow users to get access to Student data for more than 500 students, which is the hard limit in the web-interface.  So giving them access to a CSV with all filtered student data should help with that.
## HOW
- Styling for the front-end was already in place and just commented out until we were ready, so I uncommented it.  I also used the Data Export front-end code as a template to add new functions to handle download requests.
- Add a simple action that takes the `current_user.id` and enqueues the worker to send the email report to that user
- Add an Enqueue worker that takes a `user_id` and looks up that user's saved filters so that we can build CSVs that match those selected filters.  This worker also provided default filter values in cases where there are no saved values.
- Add a Send worker that takes the filter data from the Enqueue worker, uses it to run all the various Admin Diagnostic queries and combine the data in cases where a single report uses multiple API endpoints.  Then feed that data through the CSV data adapters to get CSV versions of the report, upload those CSVs to S3, and then send out an email to the user with links to those CSVs
- Add query adapters to convert each Admin Diagnostic tab (Overview, Skills, and Students) into CSVs
- Add a Mailer that takes links to the CSVs and provides them to the requesting user

### Notion Card Links
https://www.notion.so/quill/CSV-layout-for-Admin-Diagnostic-Growth-Report-384848b258eb41f3979e24c5a76f52b4#09a27a1d9c7d480aa0485ba770773fdf

### What have you done to QA this feature?
- On staging, with two different users (ids `4226515` and `5785736`), test that when a user requests a Download, that a CSV is actually generated and that it matches the web version of the reports under the following conditions:
- - No filters selected
- - Filtering by grade, teacher, and classroom
- - Selecting non-default "Group by" values
- - Selecting different "Group by" values on the Overview page vs the Skills page
- - Selecting non-default "Diagnostic" values
- - Selecting different "Diagnostic" values on the Skills page vs the Students page
- Confirm that when a user is on the Students tab and has clicked "load more" until 500 students are showing, that the helper link to download a CSV report for more data appears, and that clicking it successfully downloads the data.
- Confirm that the number of records in the Student CSV has a number of records equal to the total number of students relevant to the selected filters (rather than maxing out at 500 like the web interface does)


PR Checklist | Your Answer
------------ | -------------
Have you added and/or updated tests? |  Yes
Have you deployed to Staging? | Yes
Self-Review: Have you done an initial self-review of the code below on Github? | Yes

eadams17 committed:
add download report link

eadams17 committed:
tweak logic

eadams17 committed:
fix tests

anathomical committed:
Stub out basic email to prove end-to-end functionality

anathomical committed:
Try wrapping multiple URLs in p tags

anathomical committed:
Update skills columns and add students CSV

anathomical committed:
Fix syntax for cell formatting

anathomical committed:
Finish basic version of Overview compostition

anathomical committed:
Fix formatting

anathomical committed:
Fix formatting of cells

anathomical committed:
Convert decimal percentages to integer percentages

anathomical committed:
Update student formatting for scores

anathomical committed:
Fix typo

anathomical committed:
Weight rollup percentages for student reports

We actually calculate this on the front-end weighted before displaying it to users rather than using the API data, so update the API to return the matching data

anathomical committed:
Add controller and worker to enqueue CSV emails

anathomical committed:
Extract filters from their JSON column

anathomical committed:
Handle proper data shape

anathomical committed:
Refactor filter extraction to handle more cases

anathomical committed:
Map default schools list to ids

anathomical committed:
Update front-end for downloads

anathomical committed:
Fix import path

anathomical committed:
Fix imports

anathomical committed:
Refactor to use individual tab saved params

anathomical committed:
Fix multi-tab config

anathomical committed:
Fix fallback logic for missing stored filters

anathomical committed:
Extract correct params

anathomical committed:
Temp change for testing

anathomical committed:
Enqueue asynchronously

anathomical committed:
Merge branch 'develop' of github.com:empirical-org/Empirical-Core into lms/admin-diagnostic-csv-download

anathomical committed:
Add controller spec for new action

anathomical committed:
Update snapshot for reports page with new download UX

anathomical committed:
Add specs for CSV adapters

anathomical committed:
Update CSV formatting with tweaks from Peter S

anathomical committed:
Update "tooltip" row data with custom copy

anathomical committed:
Refactor "download" function hierarchy

So that specific functions are defined as close to their call point as possible rather than being passed in from containers that don't use them anywhere else

anathomical committed:
Merge remote-tracking branch 'origin/agdr/max-download-link' into lms/admin-diagnostic-csv-download

anathomical committed:
Update Student 500-results "download" button to trigger CSV download

anathomical committed:
Add a spec for the enqueue worker

anathomical committed:
Update worker with some bug fixes

anathomical committed:
Get initial spec into working state for email sender

anathomical committed:
WIP spec for email send

anathomical committed:
Finish spec for send email worker

anathomical committed:
Update styling for Download button on ADGR Student tab

anathomical committed:
Fix backend linting

anathomical committed:
Fix front-end linting errors

anathomical committed:
More front-end linting fixes

anathomical committed:
Fix some linting errors

anathomical committed:
Fix more linting errors

anathomical committed:
Fix front-end code

anathomical committed:
Refactor to reduce cyclomatic complexity

anathomical committed:
Update class definitions for front-end pseudo-button

anathomical committed:
Synchronously run download script for debugging purposes

anathomical committed:
Enqueue job should actually enqueue

anathomical committed:
Chasing down why this crashes in staging but not local

anathomical committed:
Still trying to figure out this staging failure

anathomical committed:
Fix cylcomatic complexity linting errors

anathomical committed:
Try symbolizing keys

anathomical committed:
Use nil-safe syntax

anathomical committed:
Remove puts call

anathomical committed:
Try fixing front-end linting error

anathomical committed:
Temporarily send download emails to Peter S for testing

anathomical committed:
Update specs with symbolize_keys expectations

anathomical committed:
Send sample CSV to me and Peter S

anathomical committed:
Make sure not to LIMIT Student report data in the CSV

anathomical committed:
Fix Student tab diagnostic selection drop-down

Looks like something got broken during a merge

anathomical committed:
Don't limit in Ruby when we're not limiting in SQL

anathomical committed:
Update the way we allow non-LIMITED student data requests

anathomical committed:
Fix filter pass-through

anathomical committed:
Use string keys instead of symbols

anathomical committed:
Do some tweaks to safely handle Overviews with no assigned students

anathomical committed:
Make sure that there are fallback keys for recommendations

anathomical committed:
Fix linting

anathomical committed:
Actually use requesting user's email for the report

Instead of the testing we were doing sending things directly to me

anathomical committed:
Merge branch 'develop' of github.com:empirical-org/Empirical-Core into lms/admin-diagnostic-csv-download

happythenewsad commented on services/QuillLMS/app/workers/admin_diagnostic_reports/enqueue_csv_email_worker.rb:
```diff
@@ -0,0 +1,70 @@
+    DEFAULT_AGGREGATION = 'grade'
+    DEFAULT_DIAGNOSTIC_ID = 1663
+
+    def perform(user_id)
```
I haven't finished reviewing, but my initial question is: why break the async work into two workers? I get that both workers are doing a lot of stuff, but that complexity can be managed with plain Ruby classes, rather than multiple workers. 

Related: If we're keeping two workers, the name should be more expansive. As in,  `PreprocessFiltersAndEnqueue*` vs. `Enqueue*`

anathomical replied:
This is a fair question.  I don't think it has to be split into two parts, but I'm partially anticipating a future feature (which may be a mistake at this point).  Specifically, the plan is to eventually add the ability for admins to subscribe to regular emails of these reports.  I expect that if we do that, then we'll actually be using a different set of saved filters, so separating the logic of extracting and formatting filters from the actual CSV generation made sense to me.  Open to the idea that we should just wait until building that future feature before worrying about the split.

happythenewsad replied:
I think it's best to not split it for now. Reasoning:

- Outside of the context of this PR diff, the intention of the code is harder to follow when it's split across two jobs
- every job adds additional overhead of managing idempotency and failure/retry states

dandrabik commented on services/QuillLMS/app/queries/admin_diagnostic_reports/diagnostic_performance_by_student_view_query.rb:
```diff
@@ -9,10 +9,11 @@ class InvalidDiagnosticIdError < StandardError; end
       raise InvalidDiagnosticIdError, "#{diagnostic_id} is not a valid diagnostic_id value." unless DIAGNOSTIC_ORDER_BY_ID.include?(diagnostic_id.to_i)
 
       @diagnostic_id = diagnostic_id
+      @limited = limited
```
You should add an `attr_reader :limited` so you can reference it without the `@`. It's added protection against accidentally changing the var.

dandrabik commented on services/QuillLMS/app/queries/admin_diagnostic_reports/diagnostic_performance_by_student_view_query.rb:
```diff
@@ -81,7 +82,7 @@ def group_by_clause
 
     def order_by_clause = "ORDER BY TRIM(SUBSTR(TRIM(student_name), STRPOS(student_name, ' ') + 1)), student_name, student_id, skill_group_name"
-    def limit_clause = " LIMIT 5000"
+    def limit_clause = " LIMIT 5000" if @limited
```
e.g. `if limited`

dandrabik commented on services/QuillLMS/app/controllers/admin_diagnostic_reports_controller.rb:
```diff
@@ -3,17 +3,23 @@
   end
 
+  def download
+    WORKERS_FOR_ACTIONS[action_name].perform_async(current_user.id)
```
This might just be better as: ` AdminDiagnosticReports::EnqueueCsvEmailWorker.perform_async(current_user.id)`. I'm not sure much is gained by this abstraction since each endpoint is custom code.

dandrabik commented on services/QuillLMS/app/queries/adapters/csv/admin_diagnostic_overview_data_export.rb:
```diff
@@ -0,0 +1,73 @@
+      class UnhandledColumnError < StandardError; end
+      class BigQueryResultMissingRequestedColumnError < StandardError; end
+
+      def self.ordered_columns
```
Is there are reason these are all class variables? This would fit well into a Service Object pattern, e.g. You could create an `ApplicationService` that returns a csv. The `to_csv_string` would be the `run` method. I think that would make the code intentions clearer.

dandrabik replied:
Also, you might consider making this a constant even if you reference as a var, e.g.
```ruby
ORDER_COLUMNS = {....}

def ordered_columns = ORDERED_COLUMNS

brendanshean replied:
I would agree with this point.  I found the class variables surprising, especially in the inheritance hierarchy.

anathomical replied:
I wrote these "adapters" following an established pattern from `app/queries/adapters/csv/admin_premium_data_export.rb` that we deployed last year for the premium data export report.  I'm not opposed to refactoring them into a service, but figured that there was some value to maintaining consistency of approach.  If we do refactor these, I imagine that we'd also want to refactor the historical example, too, for consistency sake.

brendanshean replied:
I think refactoring them would improve the readability of the code.  Initially there were just a couple of class methods on that file, but now there are a bunch.  I should have called that out in the earlier PR, so my mistake there.

dandrabik commented on services/QuillLMS/app/queries/adapters/csv/admin_diagnostic_skills_summary_data_export.rb:
```diff
@@ -0,0 +1,63 @@
+        }
+      end
+
+      def self.format_all_caps = ->(x) { x.upcase }
```
Seems like a lot of this are repeated. Maybe they should be in the base class so they don't drift from each other.

Alternatively, I think you might want to break off a separate Service Object for the formatter (general code structure), e.g.

```ruby
class Formatter < ApplicationService

  FORMATTERS = {
    all_caps: ->(x) { x.upcase },
    rounded_integer: ->(x) { x.present? ? x.round : format_blank_as_zero.call(x) },
    ...
  }

  def initialize(format, value)
    ...
  end

  def run = FORMATTERS[format]&.call(value)

end

## Then you can use
def formats
  {
   skill_group_name: :all_caps,
  }
end

Formatter.run(format, value)

dandrabik commented on services/QuillLMS/app/queries/adapters/csv/csv_data_export.rb:
```diff
@@ -0,0 +1,65 @@
+        validate_input!(processed_result, sym_columns)
+
+        CSV.generate do |csv|
+          csv << human_displayable_csv_headers(sym_columns)
```
Do we need `human_displayable` as part of the var names? Would `csv_headers` and `csv_tooltips` be clear enough?

dandrabik commented on services/QuillLMS/app/queries/adapters/csv/csv_data_export.rb:
```diff
@@ -0,0 +1,65 @@
+      def self.human_displayable_csv_tooltips(sym_columns) = sym_columns.map{|col| ordered_columns[col][1] }
+
+      def self.format_cell(sym_column, value)
+        return (format_lambdas[sym_column].call(value)) if format_lambdas.include?(sym_column)
```
Take or leave, but you might consider making a `default` formatter which does nothing but return the value, and require the code the specify the value of each column (even if it's `default`). That extra explicitness might be a nice check that you've made a formatting decision about every column and then you could remove this `if`.

anathomical replied:
I like this idea a lot, and have implemented it.

dandrabik commented on services/QuillLMS/app/workers/admin_diagnostic_reports/enqueue_csv_email_worker.rb:
```diff
@@ -0,0 +1,70 @@
+      processed_filters
+    end
+
+    private def overview_filters
```
Could these all be methods on the `AdminReportFilterSelection` instance?

anathomical replied:
Good call.

dandrabik commented on services/QuillLMS/app/workers/admin_diagnostic_reports/send_csv_email_worker.rb:
```diff
@@ -0,0 +1,146 @@
+      }
+    end
+
+    private def overview_link
```
A lot of these submethods are pretty complicated, e.g. this one parses filters, runs 5 queries, does 4 merges of data, generates a csv string, and then uploads that csv string. This could be its own service object (possibly several). I wonder if you sliced this code in verticals by report, you might be able to break it into smaller, one responsibility services, i.e. instead of generating all of the filters for the reports in one job, and then generating all of the reports in one job, you could have, for `one` report, a service to generate the filters, a service to run the queries and generate the data in the right shape, a service to generate the csv (you have this more or less) and upload the csv (this could be a generic service). I think you'd end up with much smaller classes that have one clear responsibility. It doesn't have to be that exact setup, but reading this code, each method seems like a service unto itself.

anathomical replied:
Refactored into a number of smaller services for clarity.

dandrabik reviewed (changes requested):
Nice work on getting this running. I left several refactor comments that I think it's worth do a refactor pass on this to make sure this is maintainable long term. I think you've done a good job getting the logic and functionality working, and I think the code would now benefit from some reorganization. I'm happy to jump on a call to chat through any of these comments.

anathomical committed:
Use attr_reader instead of referencing instance variables directly

anathomical committed:
Remove unnecessary abstraction

So that it is more obvious what the download action does

anathomical committed:
Move column formatting logic for CSVs into its own Service

anathomical committed:
Add a spec file for new Formatter service

anathomical committed:
Move column definitions into constants

Mostly stylistic, but hopefully makes it more obvious that these are effective configurations

anathomical committed:
Move aggregate and diagnostic_id extraction to model

And refactor worker extensively to use the now-simplified data access

anathomical committed:
Break various parts of the code into services

anathomical committed:
Fix linting errors

anathomical committed:
Make sure not to LIMIT 500 when building CSV Student reports

happythenewsad commented on services/QuillLMS/app/mailers/admin_diagnostic_reports/report_mailer.rb:
```diff
@@ -0,0 +1,21 @@
+      @human_date = DateTime.current.strftime('%B %d, %Y')
+
+      #email = user.email
+      email = 'thomas@quill.org'
```
Better address these two lines before production :-)

happythenewsad replied:
In the past, I've used an app setting, so that I can mail myself things during development, but also have the same code deployable to prod.

anathomical replied:
I've followed your pattern.

happythenewsad commented on services/QuillLMS/app/queries/adapters/csv/csv_data_export.rb:
```diff
@@ -0,0 +1,58 @@
+# frozen_string_literal: true
```
This class shares a lot of similar method implementations and signatures to AdminPremiumDataExport. 

- `row_contains_requested_columns?`, identical
- `BigQueryResultMissingRequestedColumnError`, identical 
- `format_cell`, `to_csv_string`, `validate_input!` similar 

Please consider refactoring this class so that code is less duplicated, via inheritance, module inclusion, or some other method.

anathomical replied:
I've refactored this adapter, and also refactored your upload worker to use the new `UploadToS3` service.

anathomical committed:
Refactor AdminPremiumDataExport adapter to use new base class

anathomical committed:
Make sure to freeze iterable constants

anathomical committed:
Add specs for new formatters

anathomical committed:
Remove no-longer-used workers

In favor of a simple, unified worker

anathomical committed:
Refactor Snapshots::PremiumDownloadReportsWorker to use UploadToS3

Since we've refactored that logic into a service and don't need to duplicate it in the worker anymore

anathomical committed:
Add spec for UploadToS3 service

anathomical committed:
Add overview assembler spec

anathomical committed:
Add specs for Skills and Students assemblers

anathomical committed:
Add specs for payload construction services

anathomical committed:
Merge branch 'develop' of github.com:empirical-org/Empirical-Core into lms/admin-diagnostic-csv-download

anathomical committed:
Fix linting errors

anathomical committed:
More linting fixes

anathomical committed:
Allow ENV-based email recipient overrides for testing in staging

anathomical committed:
Update controller spec for new workflow

This one removes the old Enqueue job

anathomical committed:
Update snapshot

anathomical committed:
Update Send spec for new workflow

anathomical committed:
Fix trailing white space

brendanshean commented on services/QuillLMS/app/mailers/admin_diagnostic_reports/report_mailer.rb:
```diff
@@ -0,0 +1,20 @@
+
+module AdminDiagnosticReports
+  class ReportMailer < ::UserMailer
+    default from: "The Quill Team <hello@quill.org>"
```
`default from:` is already inherited from `::UserMailer`

brendanshean commented on services/QuillLMS/app/controllers/admin_diagnostic_reports_controller.rb:
```diff
@@ -6,14 +6,24 @@ class AdminDiagnosticReportsController < ApplicationController
+
+  before_action :set_query, only: [:report]
+  before_action :validate_request, only: [:report]
+  before_action :authorize_request, only: [:report]
```
Does download action need to be authorized?

anathomical replied:
Good question.  I think that in this case, because the code logic generates and sends a report to `current_user`, we don't need auth.  You can never get access to data that's not yours.  Open to discussion, though.

brendanshean commented on services/QuillLMS/spec/services/admin_diagnostic_reports/construct_overview_query_payload_spec.rb:
```diff
@@ -0,0 +1,25 @@
+    context 'no saved specific filters' do
+      let(:specific_cilters) { nil }
+
+      it { expect(described_class::DEFAULT_AGGREGATION).to eq(subject[:aggregation]) }
```
`subject` should be in the expect clause:

```
      it { expect(subject[:aggregation]).to eq(described_class::DEFAULT_AGGREGATION) }
```

brendanshean commented on services/QuillLMS/app/services/upload_to_s3.rb:
```diff
@@ -0,0 +1,33 @@
+# frozen_string_literal: true
+
+class UploadToS3 < ApplicationService
```
I really like this extraction.  I think it needs to be renamed to `UploadCsvToS3` or something with CSV since it's not a generic uploader but a `CSV` one.

brendanshean commented on services/QuillLMS/app/services/admin_diagnostic_reports/construct_query_payload.rb:
```diff
@@ -0,0 +1,59 @@
+      @specific_filters_name = specific_filters_name
+    end
+
+    def run
```
`def run = payload`

happythenewsad replied:
This comment should be tagged [Optional], since it's not currently enforced by our Rubocop or our style guide. 

Rubocop's style guide considers it "opt-in": https://github.com/rubocop/ruby-style-guide?tab=readme-ov-file#endless-methods

There's also a Cop available: https://www.rubydoc.info/gems/rubocop/RuboCop/Cop/Style/EndlessMethod. We have the Cop enabled in LMS, but not to the specificity that it *requires* endless methods.

FWIW I like the Endless Method syntax in most cases, but want to allow for developer discretion in cases we don't necessarily anticipate right now.

brendanshean replied:
As we've discussed in engineering meetings, the endless syntax is **always** optional.  I only point it out as a stylistic suggestion.

brendanshean replied:
But I think your point about explicitly tagging as optional is correct.

brendanshean commented on services/QuillLMS/app/queries/admin_diagnostic_reports/diagnostic_recommendations_by_student_query.rb:
```diff
@@ -51,7 +52,7 @@ def relevant_diagnostic_where_clause = "AND activity_id = #{diagnostic_id}"
     def group_by_clause = ""
     def order_by_clause = "ORDER BY TRIM(SUBSTR(TRIM(students.name), STRPOS(students.name, ' ') + 1))"
-    def limit_clause = "LIMIT 500"
+    def limit_clause = "LIMIT 500" if @limited
```
Does this need attr_reader?

anathomical replied:
Yes.  Good catch.  I thought I'd cleaned all those up, but missed this one.

brendanshean commented on services/QuillLMS/spec/services/formatter_spec.rb:
```diff
@@ -0,0 +1,138 @@
+    let(:selected_format) { described_class::DEFAULT }
+    let(:value) { 'Should not be modified' }
+
+    it { expect(subject).to eq(value) }
```
The spec file switches between `it { expect(subject)... }` and `it { is_expected }`. For consistency, what about using `it { is_expected... }` throughout?

brendanshean reviewed (commented):
I've left some general comments.  I think the CSV Data Export piece should use a Service Object pattern, but short of that, I think this is good to go. 

Overall, this is quite an impressive PR pulling in a lot of moving parts.

happythenewsad reviewed (approved):
Big lift 🏋️‍♂️ 

Thanks for doing the refactors

anathomical committed:
Make changes based on code review

anathomical committed:
Refactor UploadToS3 to be a bit more generic

So that we can actually use it to upload non-CSV files at a later date if we want to

brendanshean commented on services/QuillLMS/app/services/upload_to_s3.rb:
```diff
@@ -0,0 +1,47 @@
+    # store! returns nil on failure, rather than raising an exception.
+    # We address this gotcha by manually raising an exception.
+    upload_status = uploader.store!(local_tempfile)
+    raise CloudUploadError, "Unable to upload CSV for user #{user.id}" unless upload_status
```
The extension should be extracted out of this string: `Unable to upload #{extension} for user #{user.id}`

brendanshean commented on services/QuillLMS/app/services/upload_to_s3.rb:
```diff
@@ -0,0 +1,47 @@
+  def run
+    local_tempfile << payload
+
+    upload_csv
```
Maybe just `upload` (or `upload_file`) here?

brendanshean reviewed (commented):
A couple more comments on UploadToCSV

dandrabik reviewed (approved):
This looks much cleaner with more focused classes. Thanks for cleaning this up. Looks good to me.

anathomical committed:
Move CSV generation logic over to services

anathomical committed:
Replace Adapters::Csv with specific Services

To better follow the services pattern

anathomical committed:
Clean up references to CSV from uploader service

To reflect its more generic nature

anathomical committed:
Remove specs covering removed files

brendanshean reviewed (approved):
Thank you for doing the service object extraction.  This looks great! :shipit:

anathomical committed:
Make sure to specify required keys for some queries

These were not erroring in testing earlier, but are now.  I assume a data change has revealed an unhandled edge case

anathomical committed:
Use named function args when required
